modules/collect/src/main/java/com/opengamma/strata/collect/Messages.java

```diff
@@ -110,4 +118,74 @@ public static String format(String messageTemplate, Object... args) {
     return builder.toString();
   }
 
+  /**
+   * Formats a templated message inserting named arguments.
+   * <p>
+   * This method combines a template message with a list of specific arguments.
+   * It can be useful to delay string concatenation, which is sometimes a performance issue.
+   * The approach is similar to SLF4J MessageFormat, Guava Preconditions and String format().
+   * <p>
+   * The message template contains zero to many "{name}" placeholders.
+   * Each placeholder is replaced by the next available argument.
+   * Empty "{}" placeholders do not get replaced, and will be present in the output in the same form.
+   * If there are too few arguments, then an {@link IllegalArgumentException} will be thrown.
+   * If there are too many arguments, then the excess arguments are ignored.
+   * No attempt is made to format the arguments.
+   * <p>
+   * This method is null tolerant to ensure that use in exception construction will
+   * not throw another exception, which might hide the intended exception.
+   *
+   * @param messageTemplate  the message template with "{}" placeholders, null returns empty string
+   * @param args  the message arguments, null treated as empty array
+   * @return the formatted message
+   * @throws IllegalArgumentException if you have provided less arguments than placeholders
+   */
+  public static Pair<String, Map<String, String>> formatWithAttributes(String messageTemplate, Object... args) {
+    if (messageTemplate == null) {
+      return formatWithAttributes("", args);
+    }
+    if (args == null) {
+      return formatWithAttributes(messageTemplate, new Object[0]);
+    }
+    // try to make builder big enough for the message and the args
+
+    StringBuilder outputStringBuilder = new StringBuilder(messageTemplate.length() + args.length * 20);
+    ImmutableMap.Builder<String, String> attributesMap = ImmutableMap.builder();
+
+    String pattern = "\\{(\\w+)\\}";
+    Pattern regexPattern = Pattern.compile(pattern);
+    Matcher matcher = regexPattern.matcher(messageTemplate);
```

Comment:
Using a regex might be a performance problem. But manually parsing is likely to be fiddly. I'd use a regex for simplicity until it's obviously a problem.